syntax = "proto3";

package kvstore;

message Empty {

}

message GetRequest {
  string key = 1;
}

message GetResponse {
  string value = 1;
  bool found = 2;
  bool success = 3;
  uint64 version = 4;  // Added for conflict resolution
}

message PutRequest {
  string key = 1;
  string value = 2;
  uint64 version = 3;  // Added versioning for conflict resolution
  bool is_forward = 4;  // if is forwarded by replica
}

message PutResponse {
  string old_value = 1;
  bool old_value_found = 2;
  bool success = 3;
  uint64 version = 4;  // Return version for the key
}

message DeleteRequest {
  string key = 1;
}

message DeleteResponse {
  bool success = 1;
}

message BatchGetRequest {
  repeated string keys = 1;
}

message BatchGetResponse {
  map<string, GetResponse> results = 1;
}

message BatchPutRequest {
  map<string, string> key_values = 1;
}

message BatchPutResponse {
  map<string, PutResponse> results = 1;
}

// New message types for the Die functionality
message DieRequest {
  bool clean = 1; // Added clean field to indicate the shutdown type
  string message = 2; // Optional message to send during die
}

message DieResponse {
  bool success = 1;
  string message = 2; // Optional message on die
}

// New message types for the Leave functionality
message LeaveRequest {
  bool clean = 1; // Added clean field to indicate the shutdown type
  string message = 2; // Optional message to send during die
}

message LeaveResponse {
  bool success = 1;
  string message = 2; // Optional message on die
}

// New message types for the Ping functionality
message PingRequest {
}

message PingResponse {
  bool is_alive = 1; // if the master is alive
}

message TransferToNewTailRequest {
  int32 new_tail_port = 1; // port of the new tail
  string hostname = 2; // hostname of the new tail
}
message TransferToNewTailResponse {
  bool success = 1;
}
message PromoteToHeadRequest {
}
message PromoteToHeadResponse {
  bool success = 1;
}
message TailUpdated{
  int32 new_tail_port = 1;
}
service KVStore {
  rpc Get(GetRequest) returns (GetResponse);
  rpc Put(PutRequest) returns (PutResponse);
  rpc Delete(DeleteRequest) returns (DeleteResponse);
  rpc BatchGet(BatchGetRequest) returns (BatchGetResponse);
  rpc BatchPut(BatchPutRequest) returns (BatchPutResponse);
  rpc PutToNext(PutRequest) returns (PutResponse);  // To forward Put requests to the next node in the chain
  // New Die RPC method
  rpc Die(DieRequest) returns (DieResponse);
  rpc Leave(LeaveRequest) returns (LeaveResponse);
  
  // New Ping RPC method
  rpc Ping(PingRequest) returns (PingResponse);
  rpc TransferToNewTail(TransferToNewTailRequest) returns (TransferToNewTailResponse);
  rpc PromoteToHead(PromoteToHeadRequest) returns (PromoteToHeadResponse);
  rpc ForwardToNext(PutRequest) returns (PutResponse);
}


message GetHeadRequest{
  bool replace = 1;
}
message GetTailRequest{
  bool replace = 1;
}

message GetReplicaResponse{
  int32 port = 1;
  string hostname = 2; // not needed for p2 running on localhost
  bool success = 3;
}

message SendHeartBeatRequest{
  bool is_alive = 1; // whether the child is alive
  int32 port = 2; // port of the child
}
message SendHeartBeatResponse{
  bool is_alive = 1;
}

message GetNextInChainRequest{
  int32 port = 1; // port of the current node
  string hostname = 2; // hostname of the current node
}

message StartServerRequest{
  int32 port = 1;
  string hostname = 2;
  bool new = 3;
}

message StartServerResponse{
  bool success = 1;
}

service MasterNode {
  rpc GetHead(GetHeadRequest) returns (GetReplicaResponse);
  rpc GetTail(GetTailRequest) returns (GetReplicaResponse);
  rpc SendHeartBeat(SendHeartBeatRequest) returns (SendHeartBeatResponse);
  rpc GetNextInChain(GetNextInChainRequest) returns (GetReplicaResponse);
  rpc TransferToNewTailDone(TailUpdated) returns (Empty);
  rpc StartServer(StartServerRequest) returns (StartServerResponse);
}
